odoo/tools/profiler.py

```diff
@@ -281,6 +282,37 @@ def enabled(cls):
         qweb_hooks = getattr(current_thread, 'qweb_hooks', ())
         return bool(execution_context_enabled or qweb_hooks)
 
+    @classmethod
+    def wrap_compile(cls, method_compile):
+        @functools.wraps(method_compile)
+        def _tracked_compile(irQweb, template, options):
+            if QwebTracker.enabled():
+                render_template = method_compile(irQweb, template, options)
+                def profiled_method_compile(irQweb, values):
+                    ref = options.get('ref')
+                    ref_xml = options.get('ref_xml')
+                    qweb_tracker = QwebTracker(ref, ref_xml, irQweb.env.cr)
+                    irQweb = irQweb.with_context(qweb_tracker=qweb_tracker)
+                    if qweb_tracker.execution_context_enabled:
+                        with ExecutionContext(template=ref):
+                            return render_template(irQweb, values)
+                    return render_template(irQweb, values)
+                return profiled_method_compile
+            return method_compile(irQweb, template, options)
+        return _tracked_compile
+
+    @classmethod
+    def wrap_compile_directive(cls, method_compile_directive):
+        @functools.wraps(method_compile_directive)
+        def _tracked_compile_directive(irQweb, el, options, directive, indent):
+            if QwebTracker.enabled() and directive not in ('content', 'tag'):
+                enter = irQweb._indent(f"self.env.context['qweb_tracker'].enter_directive({repr(directive)}, {repr(el.attrib)}, {repr(options['last_path_node'])})", indent)
+                leave = irQweb._indent("self.env.context['qweb_tracker'].leave_directive()", indent)
+                code_directive = method_compile_directive(irQweb, el, options, directive, indent)
+                return [enter] + code_directive + [leave] if code_directive else []
+            return method_compile_directive(irQweb, el, options, directive, indent)
+        return _tracked_compile_directive
+
```

Comment:
While I love this decorator, I would define it in the qweb file instead of the QWebTracker (feature-envy, it is qweb that uses the profiling, not the other way around). I would also rename it to coin that the method is *optionally* profiled. Also (but that's my personal taste again) I prefer `[enter, *code_directive, leave]`. Also the code is quite dense, I would use [decorator](https://pypi.org/project/decorator/) and an early-return:

```py
@classmethod
@decorator.decorator
def wrap_compile_directive(cls, method_compile_directive, irQweb, el, options, directive, indent):
    if not QwebTracker.enabled() or directive in ('content', 'tag'):
        return method_compile_directive(irQweb, el, options, directive, indent)    

    enter = irQweb._indent(f"self.env.context['qweb_tracker'].enter_directive({directive!r}, {el.attrib!r}, {options['last_path_node']!r})", indent)
    leave = irQweb._indent("self.env.context['qweb_tracker'].leave_directive()", indent)
    code_directive = method_compile_directive(irQweb, el, options, directive, indent)
    return [enter, *code_directive, leave] if code_directive else []
```

Reply:
As a reminder, don't call to `QwebTracker.enabled() `inside the `compile_directive`, use the `options['profile']` instead

Reply:
Why use you "@decorator.decorator" ?